janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexHasUniquePropertyOptimizerStrategy.java

```diff
@@ -0,0 +1,146 @@
+     */
+    @Override
+    protected boolean isValidStep(HasStep<?> step) {
+        StandardJanusGraphTx tx;
+
+        try {
+            tx = (StandardJanusGraphTx) JanusGraphTraversalUtil.getTx(step.getTraversal());
+        } catch (IllegalArgumentException iaex) {
+            // The traversal needs to be bound to a graph in order to get information about the indexes
+            return false;
+        }
+
+        List<String> givenKeys = step.getHasContainers().stream()
+            .map(HasContainer::getKey).collect(Collectors.toList());
+
+        List<PredicateCondition<String, JanusGraphElement>> constraints = step.getHasContainers().stream()
+            .filter(hc -> hc.getBiPredicate() == Compare.eq)
+            .map(hc -> new PredicateCondition<>(hc.getKey(), JanusGraphPredicateUtils.convert(hc.getBiPredicate()), hc.getValue()))
+            .collect(Collectors.toList());
+        final MultiCondition<JanusGraphElement> conditions = QueryUtil.constraints2QNF(tx, constraints);
+
+        // collect all matching indexes
+        final Set<IndexType> availableIndexes = new HashSet<>();
+        ConditionUtil.traversal(conditions, condition -> {
+            if (condition instanceof PredicateCondition) {
+                final RelationType type = ((PredicateCondition<RelationType,JanusGraphElement>) condition).getKey();
+                Preconditions.checkArgument(type != null && type.isPropertyKey());
+                Iterables.addAll(availableIndexes, ((InternalRelationType) type).getKeyIndexes());
+            }
+            return true;
+        });
+
+        // select only valid unique indexes
+        return availableIndexes.stream()
+            .filter(IndexType::isCompositeIndex)
+            .map(index -> (CompositeIndexType) index)
+            .filter(cIndex -> cIndex.getCardinality() == Cardinality.SINGLE)
+            .filter(cIndex -> isIndexSatisfiedByGivenKeys(cIndex, givenKeys))
+            .count() > 0;
+    }
+
+    private boolean isIndexSatisfiedByGivenKeys(IndexType index, Collection<String> givenKeys) {
+        for (int i = 0; i < index.getFieldKeys().length; ++i) {
+            if (!givenKeys.contains(index.getFieldKeys()[i].getFieldKey().name())) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private Traversal.Admin<?,Long> generateFilter(Traversal.Admin<?,?> traversal, HasStep<?> originalStep) {
+        Traversal.Admin filterTraversal = new DefaultGraphTraversal<>(traversal.getGraph().get())
+            .addStep(
+                new GraphStep<>(traversal, Vertex.class, true)
+            );
+
+        filterTraversal.addStep(originalStep);
+        filterTraversal.addStep(new IdStep<>(traversal));
+        return filterTraversal;
+    }
+
+    private long evaluateFilter(Traversal.Admin<?,Long> filterTraversal) {
+        // negative IDs are not valid in JanusGraph. Therefore, if the filterTraversal has no result, the newly
```

Comment:
Agreed. Although it would be an `outE()`, `inE()`,... step instead of `out()`, `in()`, ... which is not as bad performance wise. But still bad :D
So I think the cleanest way to so is to create a new strategy which eliminates traversals with unsatisfiable conditions within them completely. Sounds nice to me :grin: